sdk/python/kfp/compiler/pipeline_spec_builder.py

```diff
@@ -1575,10 +1579,30 @@ def create_pipeline_spec(
         component_spec=pipeline_spec.root,
         dag_outputs=modified_pipeline_outputs_dict,
     )
+    # call _build_dag_outputs first to verify the presence of an output annotation
+    # at all, then validate that the annotation is correct with _validate_dag_output_types
+    _validate_dag_output_types(
+        dag_outputs=modified_pipeline_outputs_dict,
+        structures_component_spec=component_spec)
```

Comment:
Type checking for outputs; e.g., `return task.outputs['single_artifact']` should not be paired with `-> List[Artifacts]`